---
title: 'BCB 546 Spring 2024 Assignment: Data inspection, wrangling and visualization
  using R programming'
author: 'Author: Usamah Kabuye; ukabuye@iastate.edu'
date: "Last compiled on:`r format(Sys.time(), '%A, %d %B %Y')`"
output:
  html_document:
    theme: readable
    df_print: paged
    toc: true
    toc_depth: '6'
    toc_float: true
  word_document:
    toc: true
    toc_depth: '6'
---

# Introduction

* This assignment has two parts

    1. Data inspection and wrangling
    
    2. Data visualization

# Part 1: Data inspection and wrangling

* We are going to use two .txt data frames (df) i.e. the fang et al genotypes and snp position data files

## Packages

* Generally 'tidyverse' package 

```{r warning=F, message=F}
library(tidyverse)
```

## Importing the two .txt files into R

```{r}
fang <- read.table("fang_et_al_genotypes.txt", header = TRUE, sep = "\t")

SNPs = read.table("snp_position.txt", header = T, sep = "\t")
```

# Data inspection

* Type of each df 

```{r}
class(fang)

class(SNPs)
```

* Size of each df

```{r}
object.size(fang)

object.size(SNPs)
```

* Viewing both data frames (df)

```{r}
View(fang)

View(SNPs)
```

* Dimensions of each df

```{r}
dim(fang) # gives number of rows and columns i.e. dimensions of the df
```

```{r}
dim(SNPs)
```

* Structure of each df

```{r}
str(fang) # column names and types of variables therein
```

```{r}
str(SNPs)
```

* Summary of SNP IDs

```{r warning=F, message=F}
library(tidyverse)
summary(select(SNPs, `SNP_ID`)) # SNP df has 983 SNP IDs
```

* Unique SNPs

```{r}
unique(SNPs$Chromosome) # In the SNPs df, chromosomes are numbered from 1 through 10. Exact location of some SNPs is unknown while some SNPs appear on multiple chromosomes.
```

# Data wrangling

## SNPs df wrangling

```{r}

SNP_nocdv <- SNPs[-2] # Delete the second column from the original SNP df

view(SNP_nocdv)

create_SNP_col <- select(SNP_nocdv, c(1,2,3))

view(create_SNP_col)
```

## Maize data wrangling

```{r}

maize_fang <- filter(fang, Group %in% c('ZMMIL', 'ZMMLR', 'ZMMMR'))

genotype_maize <- maize_fang[-(1:3)] # creates a genotype file for maize without the first 3 columns

# Transforming the file

trans_maize <- t(maize_fang) %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID")

transposed_maize <- trans_maize[4:nrow(trans_maize),]

merged_Maize <- merge(create_SNP_col, transposed_maize, by = "SNP_ID")

view(merged_Maize)
```

### Create separate directory for maize files

```{r warning=F, message=F}
dir.create('./Maize_dir')
```

### Ten maize files based on SNP increasing position values

* Ten Maize files one for each chromosome with SNPs based on increasing position values and missing data encoded by "?" symbol

```{r}
chr_maize <- filter(merged_Maize, Chromosome != "multiple" & Chromosome != "unknown")

for (i in 1:length(unique(chr_maize$Chromosome))) {
  chrom <- chr_maize %>% filter(Chromosome == i) %>% arrange(Position)
  write.table (chrom, file = paste("./Maize_dir/Maize_incr", i, sep = "_"), quote = FALSE)
}
```

### Ten maize files based on SNP decreasing position values

* Ten Maize files one for each chromosome with SNPs based on decreasing position values and missing data encoded by "_" symbol

```{r}
Decr_Maize <- as_tibble(lapply(merged_Maize, gsub, pattern ="?", replacement ="-", fixed = TRUE))

Chrom_Decr_Maize <- filter(Decr_Maize, Chromosome != "multiple" & Chromosome != "unknown")
for (i in 1:length(unique(Chrom_Decr_Maize$Chromosome))) {
  chrm_decreasing <- Chrom_Decr_Maize %>% filter(Chromosome == i) %>% arrange(desc(Position))
  write.table(chrm_decreasing, file = paste("./Maize_dir/Maize_decr", i, sep = "_"), quote = FALSE)
}
```

## Teosinte data wrangling

```{r}
Teosinte_fang <- filter(fang, Group %in% c('ZMPBA', 'ZMPIL', 'ZMPJA'))

genotype_teosinte <- Teosinte_fang[-(1:3)] # create a genotype file for teosinte without the first 3 columns

# Transforming the df

trans_teosinte <- t(Teosinte_fang) %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID")

view(trans_teosinte)

transposed_teosinte <- trans_teosinte[4:nrow(trans_teosinte),]

view(transposed_teosinte)

merged_teosinte <- merge(create_SNP_col, transposed_teosinte, by = "SNP_ID", all.x = TRUE, all.y = TRUE)

view(merged_teosinte)
```

### Create separate directory for maize files

```{r warning=F, message=F}
dir.create('./Teosinte_dir')
```

### Ten teosinte files based on SNP increasing position values

* Ten teosinte files one for each chromosome with SNPs based on increasing position values and missing data encoded by "?" symbol

```{r}
chr_teo <- filter(merged_teosinte, Chromosome != "multiple" & Chromosome != "unknown")

for (i in 1:length(unique(chr_teo$Chromosome))) {
  Tchrom <- chr_teo %>% filter(Chromosome == i) %>% arrange(Position)
  write.table (Tchrom, file = paste("./Teosinte_dir/Teosinte_incr", i, sep = "_"), quote = FALSE)
}
```

### Ten teosinte files based on SNP decreasing position values

* Ten teosinte files one for each chromosome with SNPs based on decreasing position values and missing data encoded by "-" symbol

```{r}
Decr_Teosinte <- as_tibble(lapply(merged_teosinte, gsub, pattern ="?", replacement ="-", fixed = TRUE))

Tchrom <- filter(Decr_Teosinte, Chromosome != "multiple" & Chromosome != "unknown")
for (i in 1:length(unique(Tchrom$Chromosome))) {
  Tchrm_decreasing <- Tchrom %>% filter(Chromosome == i) %>% arrange(desc(Position))
  write.table(Tchrm_decreasing, file = paste("./Teosinte_dir/Teosinte_decr", i, sep = "_"), quote = FALSE)
}
```

# Data visualization

## Packages

* 'Tidyverse' and 'ggplot2'

```{r warning=FALSE, message=FALSE}
library(tidyverse)

library(ggplot2)
```

## Fang data frame wrangling for visualization

```{r}

visual_fang <- fang %>% select(-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID")

Genotypes_SNP <- merge(create_SNP_col, visual_fang) %>% filter(., !Chromosome %in% c("unknown", "multiple"))

view(Genotypes_SNP)
```

## Visualization of SNPs per chromosome

* Distribution of SNPs on each chromosome

```{r}
Distr_on_Chr <- 
  ggplot(Genotypes_SNP, aes(x=as.double(Chromosome), #x=as.double(Chromosome) considers Chr as a number
                            fill = as.factor(as.double(Chromosome)))) +  # for getting chrom in the right order in the legend   
  geom_bar() +
  scale_x_continuous(breaks = 1:10) + # bcs x=as.double(Chromosome), breaks by default don't match chrom number
  theme_bw() + 
  ggtitle("Distribution of SNPs per chromosome") +
  labs(x = "Chromosome", y = "Total SNPs", fill = "Chromosome") #fill = "Chromosome" only replace legend's title

Distr_on_Chr
```

## Visualization of SNPs across chromosomes

* Distribution of SNPs across each chromosome

```{r}
Distr_across_chr <- 
  ggplot(Genotypes_SNP, aes(x= as.numeric(Position))) + 
  geom_density(aes(fill = as.factor(as.double(Chromosome)))) +  #same as the previous plot 
  facet_wrap(~ as.factor(as.double(Chromosome)), nrow = 2, ncol = 5) + 
  ggtitle("Diversity of SNPs per chromosome") +
  theme(axis.text.x=element_text(angle = 90)) +                 #change orientation of x axis
  labs(x = "Position", y = "Density", fill = "Chromosome")

Distr_across_chr
```

## Compiled visualization of distribution of SNPs on and across Chromosomes

* Portable document format (pdf) print outs for visualizations of SNPs on and across chromosomes

```{r}
pdf("SNP_Visualisation.pdf")
print(Distr_on_Chr)
print(Distr_across_chr)
dev.off()
```

## Visualization of missing SNP data and amount of heterozygosity

* Proportion of homozygous and heterozygous sites in addition to missing data in each sample and each group

```{r}
tidy_fang <- fang %>% select(-JG_OTU) %>% pivot_longer( -Sample_ID: -Group, names_to = "SNP_ID", values_to = "Sequence")

tidy_fang_col <- tidy_fang %>% mutate(new_sequence = ifelse(Sequence %in% c("A/A", "T/T", "C/C", "G/G"), "Homozygous", ifelse(Sequence == "?/?", "Missing", "Heterozygous"))) 

view(tidy_fang_col)
```

## Missing SNPs and Heterozygosity

```{r}
All_Fang <-  ggplot(tidy_fang_col, aes(x = Sample_ID, fill = new_sequence)) +
  ggtitle("Heterozygosity Plot") +
  geom_bar(position = "fill") + theme_bw() + labs(x = "Sample ID", y = "Proportion")

# Heterozygous grouping
Groups <- ggplot(tidy_fang_col, aes(x = Group , fill = new_sequence)) + geom_bar(position = "fill") + 
  ggtitle("Heterozygosity by Groups") +
  theme_bw() + theme(axis.text.x = element_text(angle = 90))+ labs(y = "Proportion")
```

## Compiled visualization of Missing SNP data and heterozygosity

* Pdf output

```{r}
pdf("MissingDataAndHeterozygosity.pdf")
print(All_Fang)
print(Groups)
dev.off()
```

## My own visualization

### Preparing files

```{r warning=F, message=FALSE}

My_maize <- maize_fang %>% select(-JG_OTU, -Group) %>% column_to_rownames(., var = "Sample_ID") %>% t() %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID")

view(My_maize)

maize_and_SNP <- merge(create_SNP_col, My_maize) %>% filter(., !Chromosome %in% c("unknown", "multiple"))

view(maize_and_SNP)
```

### My visualization option

* Below I visualized the density variation of SNPs associated with homozygous A the Fang_et_algenotypes df

```{r}
new_fang <- fang[-(1:3)]
trans_fang <- t(new_fang) %>% as.data.frame() %>% rownames_to_column(., var = "SNP_ID")
genes <- select(SNPs, c(1,4,9))
fang_genes <- merge(genes, trans_fang, by = c("SNP_ID"), all.x = TRUE, all.y = TRUE)
fang_genes$Total_A <- rowSums(fang_genes[,-1] == "A/A")
single_file <- select(fang_genes, c(SNP_ID,Position,Total_A))
Density_SNP_for_AA <- ggplot(data = single_file) + geom_density(mapping = aes(x=Total_A), fill="grey")
pdf("Density_of_SNPs_associated_with_Homozygotes_of_AA.pdf")
print(Density_SNP_for_AA)
dev.off()
```
